text/0000-variadic-generics.md

```diff
@@ -0,0 +1,444 @@
+- Feature Name: variadic_generics
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow types and functions to be generic over a variable number of types.
+This will allow users to write impls which are abstract over one or more
+given types when the number of types are of an arbitrary length e.g.
+`Tuple`.
+
+# Motivation
+[motivation]: #motivation
+
+Currently implementing variadic functions is accomplished through the use
+of a variadic macro as described in the [Variadic Interfaces] section of
+the book. This may be an adequate workaround for functions, but there is
+no way to extend this to work with impls or traits. As a result, the
+following still hold:
+
+ - Implementing traits for `Tuple` could use some cleanup. For example, it
+   is currently impossible to compare two `Tuple`s with `T: Eq` where the
+   length of the tuple is above 12. The implementations of the standard
+   traits `PartialEq`, `Eq`, and `Default` are implemented in a macro
+   which is repeated for the lengths 1 - 12. See [src/libcore/tuple.rs]
+   for details.
+ - Working with function which have a variable number of arguments is
+   currently impossible. For example, a generic `bind` method,
+   `f.bind(a, b)(c) == f(a, b, c)`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Variadic generic parameters always take the form `...$ty`.
+
+### Defining types
+[defining-types]: #defining-types
+
+A `...T` in generic formal type parameters, where `...T` is the trailing
+type will define the type as variadic.
+
+```rust
+struct S<...T>(...T);
+
+S();
+S(42);
+S(12, 13, 17)
+```
+
+### Defining functions
+[defining-functions]: #defining-functions
+
+Similar to defining types the same syntax may be used in a functions generic
+parameters to define a variable length list of types.
+
+```rust
+fn bar<...T>(/* arguments */);
+```
+
+### Capturing multiple arguments
+[capturing-multiple-arguments]: #capturing-multiple-arguments
+
+In the functions arguments the pattern `...args: T` (where `T` is the name of
+the variadic list of types and `args` is the argument name) will be allowed
+to define a function which takes a variable number of arguments at the position
+of `args`.
+
+```rust
+fn bar<...T>(...args: T) -> R {...}
+
+bar();
+bar(42);
+bar(12, 13, 17);
+
+fn baz<...T>(n: i32, ...args: T) -> R { ... }
+
+baz(0);
+baz(1, 42);
+```
+
+### Unpacking variadic arguments
+[unpacking-variadic-arguments]: unpacking-variadic-arguments
+
+In order to unpack the variable length list of arguments in the function
+body, the pattern `...args` may be used in the function body.
+
+```rust
+fn f<...T>(...args: T) -> R {
```

Comment:
Instead you can just do `args: (...T)` based on the idea that `(...T)` is a tuple. This is simpler and does not break the notion of *a* formal parameter being of form `pat: type`.

Reply:
Then you would have to invoke it as `f((1, 2, 3))`, being able to specify that a function should pack extra arguments provided into a specific argument tuple is useful.

Reply:
Sure. That seems to me an ergonomics issue more than anything; `f((1,2,3))` seems fine compared to the complexity added to the language otherwise. There's also `(1, ...(2, 3, 4))` to merge tuples.